mne/chpi.py

```
@@ -419,117 +620,95 @@ def _calculate_chpi_positions(raw, t_step_min=0.1, t_step_max=10.,
     logger.info('Fitting up to %s time points (%0.1f sec duration)'
                 % (len(fit_idxs), t_end - t_begin))
     pos_0 = None
-    n_freqs = len(hpi['freqs'])
+
+    hpi['n_freqs'] = len(hpi['freqs'])
     for midpt in fit_idxs:
         #
-        # 1. Fit amplitudes for each channel from each of the N cHPI sinusoids
+        # 0. determine samples to fit.
         #
         fit_time = (midpt + raw.first_samp - hpi['n_window'] / 2.) /\
             raw.info['sfreq']
+
         time_sl = midpt - hpi['n_window'] // 2
         time_sl = slice(max(time_sl, 0),
                         min(time_sl + hpi['n_window'], len(raw.times)))
-        with use_log_level(False):
-            # loads good channels with hpi_stim
-            meg_chpi_data = raw[hpi['picks'], time_sl][0]
-        mchpi_data = np.mean(meg_chpi_data[:-1], axis=1, keepdims=True)
-        this_data = meg_chpi_data[:-1] - mchpi_data
-        chpi_data = meg_chpi_data[-1]  # copies hpi_stim
-        ons = (np.round(chpi_data).astype(np.int) &
-               hpi['on'][:, np.newaxis]).astype(bool)
-        n_on = np.sum(ons, axis=0)
-        if not (n_on >= 3).all():
-            logger.info(_time_prefix(fit_time) + '%s < 3 HPI coils turned on, '
-                        'skipping fit' % (n_on.min(),))
+
+        #
+        # 1. Fit amplitudes for each channel from each of the N cHPI sinusoids
+        #
+        sin_fit = _fit_cHPI_amplitudes(raw, time_sl, hpi, fit_time)
+
+        # skip this window if it bad.
+        # logging has already been done! Maybe turn this into an Exception
```

There are a lot of datasets for which fitting could fail briefly, so I don't think we should raise an Exception. We could make this a parameter, I guess. `bad_sin_fit='warn'` default with `'ignore' | 'raise'` options?